Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE, INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "config.h"
+#include "EnhancedSecurityTracking.h"
+
+#include <WebCore/IPAddressSpace.h>
+#include <wtf/Condition.h>
+#include <wtf/Lock.h>
+
+namespace WebKit {
+
+using namespace WebCore;
+
+static EnhancedSecuritySites& enabledSitesMap()
+{
+    ASSERT(isMainRunLoop());
+    static NeverDestroyed<EnhancedSecuritySites> staticEnabledSites;
+    return staticEnabledSites;
+}
+
+static bool didSitePreviouslyUseEnhancedSecurity(API::Navigation& navigation)
+{
+    return enabledSitesMap().contains(RegistrableDomain { navigation.currentRequest().url() });
+}
+
+void EnhancedSecurityTracking::initializeWithWebsiteDataStore(Ref<WebsiteDataStore> websiteDataStore)
+{
+    websiteDataStore->fetchEnhancedSecurityOnlyDomains([this](HashSet<RegistrableDomain>&& domains) {
+        updateEnhancedSecurityDomains(WTFMove(domains));
+    });
+}
+
+void EnhancedSecurityTracking::updateEnhancedSecurityDomains(HashSet<RegistrableDomain>&& domains)
+{
+    for (auto domain : domains)
+        enabledSitesMap().add(domain, EnhancedSecurityReason::InsecureLoad);
```

Comment:
Question: What do you want to happen if the domain is already in the map? Should we overwrite the value with `InsecureLoad`? If so, you want to call `set()`. If you call `add()` this will be a no-op if the domain is already in the map. I'm not sure what is the expected behavior but I wanted to make sure you considered this.